### PR TITLE
Add newsletter sending status page [MAILPOET-1155]

### PR DESCRIPTION
**Pull request by amine-mp**
[MAILPOET-1155]

## Note about tests

Instead of writing acceptance test for every type of newsletters. I have written generic integration tests and then tested only one newsletter type on the acceptance test. The sending status page should work the same for other types. 

[MAILPOET-1155]: https://mailpoet.atlassian.net/browse/MAILPOET-1155

### CODE REVIEW

**Comment by pavel-mailpoet** [comment]
@amine-mp I started testing your branch and I noticed straight away that when I open the new `sending-status` page it makes 2 ajax requests to the same endpoint `sending_task_subscribers` listings. I think that should be fixed. We should make only 1 request.

I am not sure if that is not my computer but those requests are quite slow. I have 12 subscribers and it takes more than 20 seconds to load the listings.

File: assets/js/src/newsletters/sending_status.jsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+      endpoint: 'newsletters',
+      action: 'get',
+      data: {
+        id: newsletterId,
+      },
+    })
+      .done(res => setNewsletterSubject(res.data.subject))
+      .fail((res) => {
+        setNewsletterSubject('');
+        MailPoet.Notice.error(
```
Comment (pavel-mailpoet) [inline assets/js/src/newsletters/sending_status.jsx]: You should use `showApiErrorNotice` instead of `error` it has some checks if everything is working.

File: assets/js/src/newsletters/sending_status.jsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        afterGetItems={(state) => {
+          MailerMixin.checkMailerStatus(state);
+          CronMixin.checkCronStatus(state);
+        }}
+      />
+    </>
+  );
+};
+SendingStatus.propTypes = {
+  location: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
```
Comment (pavel-mailpoet) [inline assets/js/src/newsletters/sending_status.jsx]: Do we always have to disable eslint on this line? you know what you need, don't you? Can't you use `shape` instead of `object` here?

File: assets/js/src/newsletters/sending_status.jsx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+}) => {
+  const resend = () => {
+    MailPoet.Ajax.post({
+      api_version: window.mailpoet_api_version,
+      endpoint: 'sending_task_subscribers',
+      action: 'resend',
+      data: { taskId, subscriberId },
+    })
+      .done(() => window.mailpoet_listing.forceUpdate())
+      .fail(res => MailPoet.Notice.error(
```
Comment (pavel-mailpoet) [inline assets/js/src/newsletters/sending_status.jsx]: Same as above, we should use `showApiErrorNotice`

File: lib/API/JSON/v1/SendingTaskSubscribers.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    $newsletter_id = !empty($data['params']['id']) ? (int)$data['params']['id'] : false;
+    $tasks_ids = SendingQueueModel::select('task_id')
+      ->where('newsletter_id', $newsletter_id)
+      ->findArray();
+    if (empty($tasks_ids)) {
+      return $this->errorResponse([
+        APIError::NOT_FOUND => __('This newsletter is not being sent to any subcriber yet.', 'mailpoet'),
+      ]);
+    }
+    $data['params']['task_ids'] = array_map(function($item) {
```
Comment (pavel-mailpoet) [inline lib/API/JSON/v1/SendingTaskSubscribers.php]: We could use `array_column` here, but since it is such a broken function, I don't know if `array_map` isn't better here...

File: tests/DataFactories/Newsletter.php [hunk 5]
```diff
@@ -213,6 +219,16 @@ public function withSendingQueue(array $options = []) {
     return $this;
   }
 
+  public function beingSentToSubscriber($subscriber, array $data = []) {
```
Comment (pavel-mailpoet) [inline tests/DataFactories/Newsletter.php]: I am not sure about the name. All other methods in the data factories start with `with`. I think this shouldn't be an exception...

File: tests/acceptance/SendingStatusCest.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      ])
+      ->create();
+    // When I visit the newsletters page
+    $I->login();
+    $I->amOnMailPoetPage('Emails');
+    $I->waitForText($newsletter->subject);
+    // I click on the "Sent to 3 of 3" link
+    $I->click('[data-automation-id="sending_status_' . $newsletter->id . '"]');
+    // Check I am on the sending status page
+    $I->seeInCurrentUrl('?page=mailpoet-newsletters#/sending-status/' . $newsletter->id);
```
Comment (pavel-mailpoet) [inline tests/acceptance/SendingStatusCest.php]: Why do we need this check?

**Comment by pavel-mailpoet** [comment]
And another thing that needs to be updated is the text of the notices, see Jira comments

File: tests/acceptance/SendingStatusCest.php (hunk 6 above)
Comment (amine-mp) [reply tests/acceptance/SendingStatusCest.php]: Just to make sure the link is working.

File: tests/DataFactories/Newsletter.php (hunk 5 above)
Comment (amine-mp) [reply tests/DataFactories/Newsletter.php]: I will rename it to `withSubscriber`.

File: tests/acceptance/SendingStatusCest.php (hunk 6 above)
Comment (pavel-mailpoet) [reply tests/acceptance/SendingStatusCest.php]: Wouldn't that be obvious from the next check(s)?

File: lib/API/JSON/v1/SendingTaskSubscribers.php (hunk 4 above)
Comment (amine-mp) [reply lib/API/JSON/v1/SendingTaskSubscribers.php]: I think `array_column` makes more sense in that case. At least it's less verbose.

File: tests/acceptance/SendingStatusCest.php (hunk 6 above)
Comment (amine-mp) [reply tests/acceptance/SendingStatusCest.php]: Yeah, you're right. I will remove that line.

File: lib/API/JSON/v1/SendingTaskSubscribers.php [hunk 7]
```diff
@@ -47,7 +47,7 @@ function listing($data = []) {
       ->findArray();
     if (empty($tasks_ids)) {
       return $this->errorResponse([
-        APIError::NOT_FOUND => __('This newsletter is not being sent to any subcriber yet.', 'mailpoet'),
+        APIError::NOT_FOUND => __( 'This email has not been sent yet.', 'mailpoet'),
```
Comment (pavel-mailpoet) [inline lib/API/JSON/v1/SendingTaskSubscribers.php]: Why did you add a space here? `__(  '`

File: assets/js/src/newsletters/sending_status.jsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  }, [newsletterId]);
+
+  return (
+    <>
+      <h1>{MailPoet.I18n.t('sendingStatusTitle')}</h1>
+      <StatsLink
+        newsletterId={newsletterId}
+        newsletterSubject={newsletterSubject}
+      />
+      {!isLoading && (
```
Comment (pavel-mailpoet) [inline assets/js/src/newsletters/sending_status.jsx]: I understand what you did here, but the User Experience is quite bad now. User will come to a blank page, sees nothing and after a while, everything just jumps to the page. Before this change, there was a table showing them we're loading the data.

File: lib/API/JSON/v1/SendingTaskSubscribers.php [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  function listing($data = []) {
+    $newsletter_id = !empty($data['params']['id']) ? (int)$data['params']['id'] : false;
+    $tasks_ids = SendingQueueModel::select('task_id')
+      ->where('newsletter_id', $newsletter_id)
+      ->findArray();
+    if (empty($tasks_ids)) {
+      return $this->errorResponse([
+        APIError::NOT_FOUND => __( 'This email has not been sent yet.', 'mailpoet'),
```
Comment (pavel-mailpoet) [inline lib/API/JSON/v1/SendingTaskSubscribers.php]: I think space shouldn't be here. `__( 'This`

File: lib/API/JSON/v1/SendingTaskSubscribers.php [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      ]);
+    }
+    $data['params']['task_ids'] = array_column($tasks_ids, 'task_id');
+    $listing_data = $this->listing_handler->get('\MailPoet\Models\ScheduledTaskSubscriber', $data);
```
Comment (pavel-mailpoet) [inline lib/API/JSON/v1/SendingTaskSubscribers.php]: Shall we use some default order? I would think that failed subscribers should be in the beginning so that it is obvious what failed.

File: lib/API/JSON/v1/SendingTaskSubscribers.php (hunk 7 above)
Comment (amine-mp) [reply lib/API/JSON/v1/SendingTaskSubscribers.php]: This was a typo.

File: lib/API/JSON/v1/SendingTaskSubscribers.php (hunk 10 above)
Comment (amine-mp) [reply lib/API/JSON/v1/SendingTaskSubscribers.php]: This may be a good idea but it's not part of the specs.

Comment (pavel-mailpoet) [reply lib/API/JSON/v1/SendingTaskSubscribers.php]: That is true, it is not part of the specs. But shouldn't we try to make the plugin the best for our users? Shouldn't we develop our features so that users would like to use them? We should go above and beyond to give our users a plugin they will love. Not just what is part of the specs.

File: assets/js/src/newsletters/sending_status.jsx (hunk 8 above)
Comment (amine-mp) [reply assets/js/src/newsletters/sending_status.jsx]: I fixed that using `React.memo`
